### PR TITLE
Feature/maskings documentation

### PR DESCRIPTION
**Pull request by fceller**
@Simran-B please check the documentation

http://jenkins01.arangodb.biz:8080/view/Documentation/job/arangodb-documentation/278/

### CODE REVIEW

**Comment by Simran-B** [comment]
@fceller "This commit has no content." ?!

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md [hunk 1]
```diff
@@ -99,12 +104,20 @@ Path
...
 
 If the path starts with a `.` then it is considered to match any path
-ending in `name`. For example, `.name` will match the attribute name
-`name` all leaf attributes in the document. Leaf attributes are
+ending in `name`. For example, `.name` will match the field
+`name` of all leaf attributes in the document. Leaf attributes are
 attributes whose value is `null` or of data type `string`, `number`,
-`bool` and `array` (see below). `name` will only match leaf attributes
+`bool` or `array`. That means, it matches `name` at the top level
+as well as at any nested level (e.g. `foo.bar.name`), but not sub-objects.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: I tested it and it seems to match up with the description. However, under _Nested objects and arrays_ there is the following statement:

> If the attribute value is an object, then the attribute is not masked.
Instead the **nested object is checked further for leaf attributes**.

If I have a document like:

```json
{
  "top": {
    "nested": {
      "deep": "foobar"
    },
  "other": "baz"
  }
}
```

Then a masking path of `top` does not seem to affect the string at `top.nested.deep`. So I'm not sure about the latter statement. Is it meant in the sense of "if _top_ is not a leaf attribute, then the next attribute at that level is checked", which would be _other_ in this case.

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md [hunk 2]
```diff
@@ -447,16 +456,17 @@ date.
...
+`begin` and `end` are in [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601)
+format.
 
-The format is described in
-[DATE_FORMAT](../../../AQL/Functions/Date.html#dateformat).
+The formatting string format is described in
+[DATE_FORMAT()](../../../AQL/Functions/Date.html#dateformat).
 
-### Integral number
+### Integral Number
```
Comment (Simran-B) [inline Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: Integral or Integer?

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md [hunk 3]
```diff
@@ -447,16 +456,17 @@ date.
...
 
 This masking type replaces the value of the attribute with a random
-integral number.  It will replace the value even if it is a string,
```
Comment (Simran-B) [inline Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: integral or integer?

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md [hunk 4]
```diff
@@ -447,16 +456,17 @@ date.
...
-boolean, or false.
+integral number. It will replace the value even if it is a string,
+Boolean, or `null`.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: false is a Boolean, so I guess this was supposed to be string, bool, null?

**Comment by Simran-B** [comment]
One more thing I noticed is that some of the error messages, e.g. if an unknown masking type is specified in the config file, then it does not tell you that the error is about maskings (or did I overlook something?). Would be nice if it always included the word "masking" somehow.

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md (hunk 1 above)
Comment (fceller) [reply Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: yes, it is meant in this way. Can you suggest a better wording?

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md (hunk 2 above)
Comment (fceller) [reply Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: Either "Integer" or "Integral Number"

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md (hunk 3 above)
Comment (fceller) [reply Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: see above

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md (hunk 4 above)
Comment (fceller) [reply Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: yes

**Comment by fceller** [comment]
Will add better error messages

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md (hunk 1 above)
Comment (Simran-B) [reply Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: If there's a loop over all attributes of an object, then I would consider that an implementation details which the user probably doesn't care about. How about the following?

```diff
-If the attribute value is an object, then the attribute is not masked.
-Instead the nested object is checked further for leaf attributes.
+If the attribute value is an object, then it is ignored and the attribute does not
+get masked. To mask nested fields, specify the full path for each leaf attribute.
```

What I actually miss feature-wise is a way to define a default behavior for nested attributes. For instance, what if you have a string attribute `foo` with a masking config to xify it. Then you change your data model and it becomes an object with the original value as a sub-attribute. If you forget to change the masking config, nothing will get masked. A pseudo path `"*"` could define the default masking behavior for the entire document (`"´*´"` for a path that is literally an asterisk), `"foo.*"` for the sub-attributes of `foo` (not sure about escaping, `"´foo.*"´` or `"foo.´*´"`?)

My idea might go too far. We can also add a simple warning that (nested) objects don't get masked unless explicitly configured and that users should verify that nothing sensitive is in the dump in the end.

**Comment by Simran-B** [comment]
Discussed the masking docs with the support team. We want to change a few more things, so please don't merge yet.

**Comment by Simran-B** [comment]
- http://jenkins01.arangodb.biz:8080/view/Documentation/job/arangodb-documentation/286/
- http://web01.arangodb.biz/feature/maskings

File: Documentation/Books/Manual/Programs/Arangodump/Maskings.md (hunk 1 above)
Comment (fceller) [reply Documentation/Books/Manual/Programs/Arangodump/Maskings.md]: Let's start with the current implementation. We can later add a flag to the definition, i.e. `maskObjects: true`.

**Comment by fceller** [comment]
push fix for default telephone number

**Comment by fceller** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2889/

**Pull request merged**